### ISSUE TITLE
Dependency lua-struct no longer maintained

### ISSUE THREAD

**Issue by jasle** [jimdigriz/mt5311#6]
The [Repository](https://github.com/iryont/lua-struct) of the dependency `lua-struct` is archived for almost a year and seems no longer maintained.

Would it be possible to replace `lua-struct`?

**Comment by jimdigriz** [comment]
I have been following and wish to support your packaging efforts in https://github.com/openwrt/packages/pull/23433. It is great to see others finding value in anyone's open source work, even more so when it is mine :)

My *personal* opinion, the OpenWRT maintainers are a little too blinded by their doctrine of "must be maintained". It misses there are some programming tasks in the world that can reach 'completion', tasks such as printing 'hello world', padding strings and (in my view) byte packing and unpacking. Others of course have differing views rooted in their own experiences. The point here is there is a line where 'maintained' is meaningless for some tasks.

There were technical reasons I picked `lua-struct`, it is 200 lines of *pure Lua code*, worked for Lua 5.1 and did not use C bindings all of which would otherwise cause problems for users when installing (and packaging!).

The focus, and arguably the title, of this issue, should be "appease OpenWRT maintainers so I can package this" as this is the actual problem. My concern is if I was to shift some goalposts and *technically* remove `lua-struct` your efforts may still be blocked by the OpenWRT maintainers.

I see a few options here:

 * use another dependency
    * if it uses C bindings, I would reject this
    * if it requires Lua 5.2 or later, I would reject this
       * unless it can be shown no one uses Lua 5.1 anymore...which would be difficult given `luajit` and `nginx`.
    * what prevents the OpenWRT maintainers stating that an alternative dependency (such as [`struct`](https://www.inf.puc-rio.br/~roberto/struct/) which is a tarball last updated 2018) is also unmaintained; are their arbitary targets "public repo" and "recently updated"?
    * maybe I missed a 'maintained' pure Lua implementation out there...patches welcomed!
 * *remove* the dependency
    * I personally enjoy violence, and so to see this move forward I would be willing just to bundle the dependency into this project which would allow you to state *technically* "we no longer rely on 'unmaintained' code"
 * amend your package so instead:
    1. depend on `lua`, `lua-posix` and `luarocks`
    2. append to your existing `install` section of your `Makefile`

            define Package/mt5211/install
                ...
                luarocks install lua-struct
                ...
            endef

Personally I prefer the last option is explored, for the reason that the origin, search engine indexing, archiving/historian processes as well as clear authorship of that part of the code base is preserved for *others* to see. Do you think this is possible and helps you make forward progress?

**Comment by jasle** [comment]
Thanks a lot for support getting this tool packaged for OpenWrt!

I have looked in the metanoia tool trying to understand how to get information from the modem and was really happy to find this tool, so I don't have to figure out and implement everything myself.

> use another dependency 

I fully understand your requirements for a potential replacement lib and did not find any promising library as a replacement too.

> remove the dependency 

Probability could solve needing `lua-strcut` as external dependency, but sounds for the reasons you named like a dirty solution.

> amend your package so instead: […]

The `define Package/mt5311/install` does run while building the package, not when installing it. So I would need to add it to the postinstall script.
But the [OpenWrt Package Contributing Guidelines](https://github.com/openwrt/packages/blob/master/CONTRIBUTING.md) also say that a package should _Have no dependencies outside the OpenWrt core packages or this repository feed_.
I would also avoid to add a dependency just for installing the real dependency bypassing the packet manager and relying on having a internet connection.

I will try to get the `mt5311` package merged together with `lua-struct` as separate package.

**Comment by jimdigriz** [comment]
> I will try to get the `mt5311` package merged together with `lua-struct` as separate package.

Let me know if I can help or if you need me for anything.

Good Luck!

**Comment by jasle** [comment]
Both packaged got merged into OpenWrt packges 🎉

Thanks a lot for your support!

I close this ticket and will open a PR for updating the Readme, when the package is available in a OpenWrt release.

**Comment by jimdigriz** [comment]
> Both packaged got merged into OpenWrt packges 🎉

Great work!

> Thanks a lot for your support!

Not a problem.

Thanks
